### PR TITLE
feat(core): add substr method to FixString class

### PR DESCRIPTION
**Pull request by dkrutskikh**
### **PR Type**
Enhancement

___

### **Description**
- Add `substr` method to `FixString` class

- Implement substring extraction with position and count parameters

- Add comprehensive test coverage for all edge cases

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["FixString class"] --> B["substr method declaration"]
  B --> C["substr implementation"]
  C --> D["comprehensive test suite"]
  D --> E["substring functionality"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>fix_string.hpp</strong><dd><code>Add substr method declaration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/fix_string.hpp

<ul><li>Add <code>substr</code> method declaration with position and count parameters<br> <li> Include comprehensive documentation with parameter descriptions<br> <li> Define method signature with default parameters and noexcept <br>specification</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/92/files#diff-2ef8f9f195db23c05d195ee56d0aaa0c714eeaf911132af27804a48fa895ce44">+20/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fix_string.inl</strong><dd><code>Implement substr method functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/fix_string.inl

<ul><li>Implement <code>substr</code> method with bounds checking and assertions<br> <li> Handle <code>npos</code> parameter for extracting to end of string<br> <li> Use <code>memcpy</code> for efficient character copying and null termination</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/92/files#diff-a47dfdd2a0becc17f40499abd90d00211a258ffaea3ce144fd5fc11d5ad08afd">+19/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>fix_string_test.cpp</strong><dd><code>Add comprehensive substr test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/core/fix_string_test.cpp

<ul><li>Add comprehensive test suite with 20+ test sections<br> <li> Cover edge cases including empty strings, boundary conditions, special <br>characters<br> <li> Test various parameter combinations and string capacities</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/92/files#diff-cc1ff39c1921067e1cd9ffcf0bc8c8f3d7349f0178f1eb6ddeac099bc0a2d2ae">+234/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

## Greptile Summary

Updated On: 2025-09-07 16:56:56 UTC

This PR adds a `substr` method to the FixString template class, which is part of the ToyGine2 game engine's core string utilities. The FixString class appears to be a fixed-capacity string implementation designed to avoid dynamic memory allocation while providing std::string-like functionality.

The implementation adds:
- A method declaration in `include/core/fix_string.hpp` with comprehensive Doxygen documentation
- The actual implementation in `include/core/fix_string.inl` that handles substring extraction
- Extensive test coverage in `test/core/fix_string_test.cpp` with 24 different test sections

The `substr` method follows standard library conventions, accepting a starting position (default 0) and count parameter (default npos), and returns a new FixString object containing the specified substring. The implementation includes proper bounds checking, handles the special `npos` value to extract to the end of the string, and maintains the same template capacity as the original string.

This addition makes FixString more complete as a string replacement by providing essential substring functionality that users would expect from a string class. The method maintains constexpr and noexcept qualifiers consistent with other FixString methods, ensuring compile-time optimization potential and exception safety.

**PR Description Notes:**
- The PR description appears to be a placeholder message about an auto-generated description that was never completed

## Important Files Changed

<details><summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| `include/core/fix_string.hpp` | 3/5 | Adds substr method declaration with comprehensive documentation following existing patterns |
| `include/core/fix_string.inl` | 4/5 | Implements substr method with proper bounds checking and edge case handling |
| `test/core/fix_string_test.cpp` | 5/5 | Adds comprehensive test coverage with 24 test sections covering all use cases and edge cases |

</details>

## Confidence score: 4/5

- This PR is generally safe to merge with minimal risk as it adds new functionality without modifying existing behavior
- Score reflects solid implementation and comprehensive testing, but slightly lowered due to minor inconsistency in assertion usage
- Pay close attention to `include/core/fix_string.inl` for the inconsistent assertion style on line 820

## Sequence Diagram
```mermaid
sequenceDiagram
    participant User
    participant FixString
    participant Memory
    participant Assert

    User->>FixString: "substr(position, count)"
    FixString->>Assert: "assert position <= _size"
    Assert-->>FixString: "validation passed"
    
    alt count == npos
        FixString->>FixString: "count = _size - position"
    end
    
    FixString->>Assert: "assert (position + count) <= _size"
    Assert-->>FixString: "validation passed"
    
    FixString->>FixString: "create result FixString"
    FixString->>FixString: "result._size = count"
    FixString->>Memory: "memcpy(result._data, _data + position, count)"
    Memory-->>FixString: "data copied"
    FixString->>FixString: "result._data[count] = '\\0'"
    FixString-->>User: "return result FixString"
```

### What change is being made?

Add a `substr` method to the `FixString` class for extracting substrings, including its implementation and comprehensive test coverage.

### Why are these changes being made?

The `substr` method enables more flexible manipulation and extraction of parts from `FixString` instances, a feature important for string processing use-cases. This addition expands the functionality of the `FixString` class and aligns with typical string class capabilities found in other libraries, promoting ease of use and extending utility.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: include/core/fix_string.hpp [hunk 1]
```diff
@@ -1652,6 +1652,26 @@ class FixString {
...
+    \return A new FixString object containing the specified substring.
+
+    \pre The \a position must be less than or equal to the string size.
+
+    \note If \a position is equal to the string size, an empty string is returned.
+    \note If \a count is \ref npos or exceeds the remaining characters from \a position, all remaining characters are
+          included.
+  */
+  constexpr inline FixString<allocatedSize> substr(std::size_t position = 0, std::size_t count = npos) const noexcept;
+
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/fix_string.hpp:1674]: fixed
